File: apps/rust-server/src/services/message_seq.rs
/// Get the next sequence number for a conversation.
/// Uses MAX(seq) + 1 to ensure monotonically increasing per-conversation.
/// Accepts any sqlx executor (pool or transaction).
pub async fn get_next_seq<'e, E>(executor: E, conversation_id: &str) -> Result<i32, sqlx::Error>
where
    E: sqlx::Executor<'e, Database = sqlx::Postgres>,
{
    let row = sqlx::query_as::<_, (Option<i32>,)>(
        r#"SELECT MAX(seq) FROM messages WHERE conversation_id = $1::uuid"#,
    )
    .bind(conversation_id)
    .fetch_one(executor)
    .await?;

    Ok(row.0.unwrap_or(0) + 1)
}

File: apps/server/src/lib/message-seq.ts
import { db } from "../db/index.js";
import { messages } from "../db/schema.js";
import { eq, sql } from "drizzle-orm";

/**
 * Get the next sequence number for a conversation.
 * Uses MAX(seq) + 1 to ensure monotonically increasing per-conversation.
 */
export async function getNextSeq(conversationId: string): Promise<number> {
  const [result] = await db
    .select({ maxSeq: sql<number>`COALESCE(MAX(${messages.seq}), 0)` })
    .from(messages)
    .where(eq(messages.conversationId, conversationId));
  return (result?.maxSeq ?? 0) + 1;
}
